src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+                || Length != o.Length)
+            {
+                throw new ArgumentException(CompilerExtensionsResources.ArgumentException_OtherNotArrayOfCorrectLength, nameof(other));
+            }
+
+            for (var i = 0; i < Length; i++)
+            {
+                var result = comparer.Compare(this[i], o[i]);
+                if (result != 0)
+                    return result;
+            }
+
+            return 0;
+        }
+
+        bool IStructuralEquatable.Equals(object other, IEqualityComparer comparer)
+        {
+            if (other is null)
+                return false;
+
+            if (!(other is SegmentedArray<T> o))
+                return false;
+
+            if ((object)_items == o._items)
+                return true;
+
+            if (Length != o.Length)
+                return false;
+
+            for (var i = 0; i < Length; i++)
+            {
+                if (!comparer.Equals(this[i], o[i]))
+                    return false;
+            }
+
+            return true;
+        }
+
+        int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
+        {
+            _ = comparer ?? throw new ArgumentNullException(nameof(comparer));
+
+            var ret = 0;
+            for (var i = Length >= 8 ? Length - 8 : 0; i < Length; i++)
+            {
+                ret = Hash.Combine(comparer.GetHashCode(this[i]), ret);
+            }
+
+            return ret;
+        }
+
+        internal TestAccessor GetTestAccessor()
+            => new TestAccessor(this);
+
+        public struct Enumerator : IEnumerator<T>
+        {
+            private readonly T[][] _items;
+            private int _nextItemSegment;
+            private int _nextItemIndex;
+            private T _current;
+
+            public Enumerator(SegmentedArray<T> array)
+            {
+                _items = array._items;
+                _nextItemSegment = 0;
+                _nextItemIndex = 0;
+                _current = default;
+            }
+
+            public T Current => _current;
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                if (_items.Length == 0)
+                    return false;
+
+                if (_nextItemIndex == _items[_nextItemSegment].Length)
+                {
+                    if (_nextItemSegment == _items.Length - 1)
+                    {
+                        return false;
+                    }
+
+                    _nextItemSegment++;
+                    _nextItemIndex = 0;
+                }
+
+                _current = _items[_nextItemSegment][_nextItemIndex];
+                _nextItemIndex++;
+                return true;
+            }
+
+            public void Reset()
+            {
+                _nextItemSegment = 0;
+                _nextItemIndex = 0;
+                _current = default;
+            }
```

Comment:
do you have tests that explicitly ensure enumeration works across a segmentedarray that is definitely built of many segments.

Reply:
No, currently only implicit.

Reply:
Added an explicit test

Reply:
`TestIterateLargeArray`